File: src/main/java/com/ultreon/bubbles/save/SavedGame.java
package com.ultreon.bubbles.save;

import org.bson.BsonDocument;
import org.bson.BsonInt64;
import org.bson.BsonString;
import org.bson.RawBsonDocument;
import org.bson.codecs.BsonDocumentCodec;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Random;

@SuppressWarnings({"unused", "FieldCanBeLocal", "SameParameterValue"})
public class SavedGame {
    private final String path;
    protected final File gameInfoFile;
    private final File directory;

    protected SavedGame(String path) {
        this(new File(path));
    }

    protected SavedGame(File directory) {
        this.path = directory.getPath();
        this.directory = directory;
        this.gameInfoFile = new File(path, "info.bson");
    }

    public static SavedGame fromFile(File file) {
        return new SavedGame(file);
    }

    public static SavedGame fromPath(Path file) {
        return new SavedGame(file.toFile());
    }

    public static SavedGame fromStringPath(String file) {
        return new SavedGame(file);
    }

    public BsonDocument loadInfoData() throws IOException {
        FileInputStream fis = new FileInputStream(gameInfoFile);
        return new RawBsonDocument(fis.readAllBytes());
    }

    public BsonDocument debugInfoData() {
        RawBsonDocument nbt = new RawBsonDocument(new BsonDocument(), new BsonDocumentCodec());
        nbt.put("name", new BsonString("Test Name - " + new Random().nextInt()));
        nbt.put("saveTime", new BsonInt64(System.currentTimeMillis()));

        return nbt;
    }

    public BsonDocument loadData(String name) throws IOException {
        return this.loadData(new File(path, name + ".bson"));
    }

    private BsonDocument loadData(File file) throws IOException {
        FileInputStream fis = new FileInputStream(file);
        return new RawBsonDocument(fis.readAllBytes());
    }

    public void dumpData(String name, BsonDocument data) throws IOException {
        this.dumpData(new File(path, name + ".bson"), data);
    }

    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private void dumpData(File file, BsonDocument data) throws IOException {
        FileOutputStream fos = new FileOutputStream(file);
        RawBsonDocument document = new RawBsonDocument(data, new BsonDocumentCodec());
        fos.write(document.getByteBuffer().array());
    }

    public File getDirectory() {
        return directory;
    }

    public boolean hasMainState() {
        return hasDataFile("game");
    }

    private boolean hasDataFile(String name) {
        return hasDataFile(new File(path, name + ".bson"));
    }

    private boolean hasDataFile(File file) {
        return file.exists() && file.getAbsolutePath().startsWith(directory.getAbsolutePath());
    }

    public void createFolders(String relPath) throws IOException {
        if (!Path.of(directory.getPath(), relPath).toFile().mkdirs()) {
            throw new IOException("Failed to create directories " + Path.of(directory.getPath(), relPath).toFile().getAbsolutePath());
        }
    }

}

File: src/main/kotlin/com/qtech/bubbles/core/common/SavedGame.kt
package com.qtech.bubbles.core.common

import net.querz.nbt.io.NBTUtil
import net.querz.nbt.tag.CompoundTag
import org.bson.BsonDocument
import org.bson.BsonInt64
import org.bson.BsonString
import org.bson.RawBsonDocument
import org.bson.codecs.BsonDocumentCodec
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*

open class SavedGame protected constructor(val directory: File) {
    private val path: String = directory.path
    protected val gameInfoFile: File = File(path, "info.bson")

    protected constructor(path: String) : this(File(path)) {}

    @Throws(IOException::class)
    fun loadInfoData(): BsonDocument {
        val fis = FileInputStream(gameInfoFile)
        return RawBsonDocument(fis.readAllBytes())
    }

    fun debugInfoData(): BsonDocument {
        val nbt = RawBsonDocument(BsonDocument(), BsonDocumentCodec())
        nbt["name"] = BsonString("Test Name - " + Random().nextInt())
        nbt["saveTime"] = BsonInt64(System.currentTimeMillis())
        return nbt
    }

    @Throws(IOException::class)
    fun loadData(name: String): CompoundTag {
        return this.loadData(File(path, "$name.dat"))
    }

    @Throws(IOException::class)
    private fun loadData(file: File): CompoundTag {
        return NBTUtil.read(file, true).tag as CompoundTag
    }

    @Throws(IOException::class)
    fun dumpData(name: String, data: CompoundTag) {
        this.dumpData(File(path, "$name.dat"), data)
    }

    @Throws(IOException::class)
    private fun dumpData(file: File, data: CompoundTag) {
        NBTUtil.write(data, file, true)
    }

    fun hasMainState(): Boolean {
        return hasDataFile("game")
    }

    @Suppress("SameParameterValue")
    private fun hasDataFile(name: String): Boolean {
        return hasDataFile(File(path, "$name.dat"))
    }

    private fun hasDataFile(file: File): Boolean {
        return file.exists() && file.absolutePath.startsWith(directory.absolutePath)
    }

    @Throws(IOException::class)
    fun createFolders(relPath: String?) {
        if (!Paths.get(directory.path, relPath).toFile().mkdirs()) {
            throw IOException("Failed to create directories " + Paths.get(directory.path, relPath).toFile().absolutePath)
        }
    }

    companion object {
        @JvmStatic
        fun fromFile(file: File): SavedGame {
            return SavedGame(file)
        }

        fun fromPath(file: Path): SavedGame {
            return SavedGame(file.toFile())
        }

        fun fromStringPath(file: String): SavedGame {
            return SavedGame(file)
        }
    }

}